Docs: add section on basic performance benchmark

sphuber opened the pull request:
The section provides a basic script to benchmark the performance of an AiiDA installation by launching a number of `ArithmeticAddCalculation` jobs. The script by default automatically sets up the required `Code` and localhost `Computer`. All created nodes are automatically deleted from the database at the end.

The documentation gives instructions on how to run the script and provides example output of a run on a typical work station including completion times for runs with variable number of daemon workers. This should give users an idea of the performance of their installation.

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -385,6 +385,53 @@ Here are a few general tips that might improve the AiiDA performance:
         and try a simple AiiDA query with the new database.
         If everything went fine, you can delete the old database location.
 
+It is not trivial to give a baseline of performance for AiiDA, as it depends a lot on the machine on which it is running, the installation, and what is being run.
```
```suggestion
The performance for AiiDA depends both on the hardware it is running on and the processes being run.
```

sphuber replied:
Wonder whether we should give more details here (or at least put it somewhere else). For example what is also important is how/where the services are running. If RabbitMQ and PostgreSQL are running on a different server, that is clearly going to incur a penalty versus them running on the localhost. Same goes for how the type of the filesystem and how it is mounted.

sphuber replied:
That makes a lot of sense. I think I will do that and already add a section in Topics. Will probably be really minimal for now, but at least we'll have a placeholder that we can add onto in the future

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -385,6 +385,53 @@ Here are a few general tips that might improve the AiiDA performance:
         If everything went fine, you can delete the old database location.
 
+It is not trivial to give a baseline of performance for AiiDA, as it depends a lot on the machine on which it is running, the installation, and what is being run.
+To still give some sense of performance, we provide this :download:`simple benchmarking script <include/scripts/performance_benchmark_base.py>` :fa:`download`.
```
```suggestion
In order to get a rough idea, you can run :download:`simple benchmarking script <include/scripts/performance_benchmark_base.py>` :fa:`download`.
```

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -385,6 +385,53 @@ Here are a few general tips that might improve the AiiDA performance:
+
+This will launch 100 ``ArithmeticAddCalculation`` jobs on the localhost and record the time until completion.
+
+Below is the output of a run using AiiDA v1.6.9 with a single daemon worker on a machine with a AMD Ryzen 5 3600 6-Core processor, with all services (RabbitMQ and PostgreSQL running on the same machine):
```
```suggestion
Below is the output for running 100 ``ArithmeticAddCalculation``s with a single daemon worker on one core of an AMD Ryzen 5 3600 6-Core processor, using AiiDA v1.6.9, and RabbitMQ and PostgreSQL running on the same machine:
```

ltalirz commented on docs/source/howto/include/scripts/performance_benchmark_base.py:
```diff
@@ -0,0 +1,127 @@
+        orm.Computer.objects.delete(computer.pk)
+        echo.echo_success(f'Deleted the created computer {computer_label}.')
+
+    echo.echo(f'Performance {(number / (time_end - time_start)):.2f} processes / s')
```
```suggestion
    echo.echo(f'Performance: {(number / (time_end - time_start)):.2f} processes / s')
```

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -385,6 +385,53 @@ Here are a few general tips that might improve the AiiDA performance:
+        Success: Deleted the created computer benchmark-5fa8c67f.
+        Performance 2.15 processes / s
+
+As you can see, it took 9.42 seconds to submit the 100 calculations.
+The entire script took 46.55 seconds to submit and successfully complete all 100 calculations.
+When running the same script and setup with a different amount of daemon workers, we got the following numbers:
```
```suggestionI
It took 9.42 seconds to submit the 100 calculations.
Overall, the entire script took 46.55 seconds to submit and successfully complete all 100 calculations, corresponding to a rate of roughly 2 calculations per second.

This rate can be sped up by adding more daemon workers:
```

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -385,6 +385,53 @@ Here are a few general tips that might improve the AiiDA performance:
+The entire script took 46.55 seconds to submit and successfully complete all 100 calculations.
+When running the same script and setup with a different amount of daemon workers, we got the following numbers:
+
+.. table::
```
This is a useful table.

If you want to see linear scaling, I guess you will need to remove the submission time.
I.e. stop the daemon, then submit the calculations, then start the daemon and measure how long the calculations take to complete (I assume this time will be roughly antiproportional to the number of daemons).

This should be straightforward with the python api you're adding for aiida v2 (but here it's indeed useful to have a v1-compatible script).

sphuber replied:
Indeed, as you say, just removing the submission time is not a fair presentation either, because as soon as the first one is submitted, the daemon will start running those. Maybe I can include a note for this explaining the processes per second stat.

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -385,6 +385,53 @@ Here are a few general tips that might improve the AiiDA performance:
+    4          16.43                   6.09
+    ========== ======================= ==========================
+
+Although there is no perfect linear scaling, an increase in the number of daemon workers clearly increases the throughput significantly.
```
```suggestion
As expected, the process rate increases with the number of daemon workers.
The scaling is not linear, since the ``ArithmeticAddCalculation`` is very quick, and submission time is not influenced by the number of workers.
```

ltalirz commented on docs/source/howto/include/scripts/performance_benchmark_base.py:
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""Basic script to benchmark the performance of an AiiDA installation."""
```
```suggestion
"""Script to benchmark the performance of the AiiDA workflow engine on a given installation.

Compatible with AiiDA version 1.6 or later.
"""
```

ltalirz reviewed (changes requested):
thanks a lot @sphuber !

I've added a few comments on the docs (haven't yet looked at the code much; I assume that will be fine)

sphuber commented:
@ltalirz I have addressed the comments and added a section to topics on performance. Even though the latter isn't really part of the focus of this PR, I already added some text as to not have it completely empty. But maybe should not go down the rabbithole too much for this PR on that part.

ltalirz commented:
I added one further commit in https://github.com/aiidateam/aiida-core/pull/5723

If that seems fine, please cherry-pick it here and close that PR

sphuber commented:
> If that seems fine, please cherry-pick it here and close that PR

Thanks, all fine. I cherry-picked the commit here. Just corrected a mistake in the script that was probably added by accident and some minor corrections in the docs (broken label).

ltalirz commented:
Thanks for the fixes!

I see I still introduced some issue
```
<string>:1: (WARNING/2) Inline interpreted text or phrase reference start-string without end-string.
/home/docs/checkouts/readthedocs.org/user_builds/aiida-core/checkouts/5635/docs/source/howto/installation.rst:341: WARNING: Inline literal start-string without end-string.
/home/docs/checkouts/readthedocs.org/user_builds/aiida-core/checkouts/5635/docs/source/topics/performance.rst:53: WARNING: download file not readable: /home/docs/checkouts/readthedocs.org/user_builds/aiida-core/checkouts/5635/docs/source/topics/include/scripts/performance_benchmark_base.py
```

sphuber commented:
> /home/docs/checkouts/readthedocs.org/user_builds/aiida-core/checkouts/5635/docs/source/topics/performance.rst:53: WARNING: download file not readable: /home/docs/checkouts/readthedocs.org/user_builds/aiida-core/checkouts/5635/docs/source/topics/include/scripts/performance_benchmark_base.py

This was just an incorrect link (it is relative, but we moved it to another subfolder).

> /home/docs/checkouts/readthedocs.org/user_builds/aiida-core/checkouts/5635/docs/source/howto/installation.rst:341: WARNING: Inline literal start-string without end-string.

This was actually a nasty one. The error messages of Sphinx are often not very useful. The problem was the following
```
``ArithmeticAddCalculation``s
```
Apparently, you cannot have this `s` straight after the closing literal block. What made it even more confusing is that this was on line *342* whereas the error message states 341 :sweat_smile: . Anyway, should be fixed now.

sphuber commented on docs/source/topics/performance.rst:
```diff
@@ -0,0 +1,67 @@
+The performance of AiiDA depends on many factors:
+
+* the hardware that AiiDA is running on
+* the environment in which AiiDA is installed
```
I had originally added this but already removed it when cherry-picking your commit, where you suggested to remove it. Think you were looking at an outdated diff.

What I mean with the environment is the runtime for AiiDA, i.e. is it installed with Conda, in virtual-env, or system wide, or maybe running in a container, etc. It is not quite clear if and how much of an impact it could have, but I thought it would be good to have the list complete. But now I think it is fine to just omit it for the time being.
